File: app/src/main/java/com/maksk993/cryptoportfolio/presentation/view/fragments/AddAssetFragment.java
package com.maksk993.cryptoportfolio.presentation.view.fragments;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import com.maksk993.cryptoportfolio.R;
import com.maksk993.cryptoportfolio.presentation.models.AssetAdapter;
import com.maksk993.cryptoportfolio.presentation.models.AssetItem;
import com.maksk993.cryptoportfolio.presentation.models.AssetViewHolder;
import com.maksk993.cryptoportfolio.presentation.models.FindFragmentById;
import com.maksk993.cryptoportfolio.presentation.viewmodel.AddAssetViewModel;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class AddAssetFragment extends Fragment {
    AddAssetViewModel viewModel;
    ImageButton btnBack;
    RecyclerView recyclerView;
    List<AssetItem> items = new ArrayList<>();
    RecyclerView.Adapter<AssetViewHolder> adapter;

    View view;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        viewModel = new ViewModelProvider(this).get(AddAssetViewModel.class);
        view = inflater.inflate(R.layout.fragment_add_asset, container, false);

        recyclerView = view.findViewById(R.id.recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        adapter = new AssetAdapter(view.getContext(), items);
        recyclerView.setAdapter(adapter);

        btnBack = view.findViewById(R.id.btn_back);
        btnBack.setOnClickListener(view -> {
            viewModel.openFragment(FindFragmentById.getInteger(FindFragmentById.PORTFOLIO));
        });

        viewModel.getPricesFromCoinMarketCap();
        viewModel.shouldNewFragmentBeOpened().observe(getViewLifecycleOwner(), new Observer<FindFragmentById>() {
            @Override
            public void onChanged(FindFragmentById findFragmentById) {
                Fragment fragment = FindFragmentById.getFragment(findFragmentById);
                FragmentTransaction transaction = requireActivity().getSupportFragmentManager().beginTransaction();
                transaction.replace(R.id.nav_host_fragment, fragment);
                transaction.commit();
            }
        });

        viewModel.getItem().observe(getViewLifecycleOwner(),
                assetItem -> updatePrices(assetItem.getName(), assetItem.getPrice())
        );

        return view;
    }

    public void updatePrices(String symbol, float price) {
        if (!items.isEmpty()) {
            for (int i = 0; i < items.size(); i++) {
                if (Objects.equals(items.get(i).getName(), symbol)) {
                    if (price != items.get(i).getPrice()) {
                        items.set(i, new AssetItem(symbol, price, R.drawable.ic_history));
                        adapter.notifyItemChanged(i);
                    }
                    return;
                }
            }
        }
        items.add(new AssetItem(symbol, price, R.drawable.ic_history));
        items.sort((a, b) -> a.getName().compareTo(b.getName()));
        int index = 0;
        for (int j = 0; j < items.size(); j++){
            if (items.get(j).getName().equals(symbol)) {
                index = j;
                break;
            }
        }
        adapter.notifyItemInserted(index);
    }
}

File: app/src/main/java/com/maksk993/cryptoportfolio/presentation/view/fragments/AddAssetFragment.kt
package com.maksk993.cryptoportfolio.presentation.view.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.LinearLayoutManager
import com.maksk993.cryptoportfolio.R
import com.maksk993.cryptoportfolio.databinding.FragmentAddAssetBinding
import com.maksk993.cryptoportfolio.presentation.models.AssetAdapter
import com.maksk993.cryptoportfolio.domain.models.AssetItem
import com.maksk993.cryptoportfolio.presentation.models.FindFragmentById
import com.maksk993.cryptoportfolio.presentation.viewmodel.MainViewModel


class AddAssetFragment : Fragment() {
    private val viewModel: MainViewModel by activityViewModels()
    private lateinit var binding: FragmentAddAssetBinding

    private lateinit var adapter : AssetAdapter
    private val items : MutableList<AssetItem> = ArrayList()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentAddAssetBinding.inflate(inflater, container, false)

        initRecyclerView()
        initButtons()
        initObservers()

        return binding.root
    }

    private fun initObservers(){
        viewModel.actualPrices.observe(viewLifecycleOwner){
            for (i in it){
                updatePrices(i.key, i.value)
            }
        }

        viewModel.updatingAsset.observe(viewLifecycleOwner){
            updatePrices(it.symbol, it.price)
        }
    }

    private fun updatePrices(symbol : String, price : Float){
        if (items.isNotEmpty()){
            for (i in 0 until items.size){
                if (items[i].symbol == symbol) {
                    if (price != items[i].price) {
                        items[i] = AssetItem(symbol, price, R.drawable.ic_money)
                        adapter.notifyItemChanged(i)
                    }
                    return
                }
            }
        }
        items.add(AssetItem(symbol, price, R.drawable.ic_money))
        items.sortWith {
            a: AssetItem, b: AssetItem ->
            a.symbol.compareTo(b.symbol)
        }
        adapter.notifyDataSetChanged()
    }

    private fun initRecyclerView(){
        binding.recycler.setLayoutManager(LinearLayoutManager(requireContext()))
        adapter = AssetAdapter(requireContext(), items)
        adapter.setOnItemClickListener { position ->
            viewModel.setFocusedAsset(items[position])
            viewModel.openFragment(FindFragmentById.INDICATE_QUANTITY)
        }
        binding.recycler.adapter = adapter
    }

    private fun initButtons(){
        binding.btnBack.setOnClickListener {
            viewModel.openFragment(FindFragmentById.PORTFOLIO)
        }
    }
}